src/Microsoft.Health.Dicom.Core/Features/Export/IdentifierExportSource.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Core.Features.Model;
+using Microsoft.Health.Dicom.Core.Features.Partition;
+using Microsoft.Health.Dicom.Core.Features.Retrieve;
+using Microsoft.Health.Dicom.Core.Messages;
+using Microsoft.Health.Dicom.Core.Models;
+using Microsoft.Health.Dicom.Core.Models.Common;
+using Microsoft.Health.Dicom.Core.Models.Export;
+
+namespace Microsoft.Health.Dicom.Core.Features.Export;
+
+internal sealed class IdentifierExportSource : IExportSource
+{
+    public event EventHandler<ReadFailureEventArgs> ReadFailure;
+
+    public TypedConfiguration<ExportSourceType> Configuration => GetConfiguration();
+
+    private readonly IInstanceStore _instanceStore;
+    private readonly PartitionEntry _partition;
+    private readonly IdentifierExportOptions _options;
+
+    private int _startIndex;
+
+    public IdentifierExportSource(IInstanceStore instanceStore, PartitionEntry partition, IOptions<IdentifierExportOptions> options)
+    {
+        _instanceStore = EnsureArg.IsNotNull(instanceStore, nameof(instanceStore));
+        _partition = EnsureArg.IsNotNull(partition, nameof(partition));
+        _options = EnsureArg.IsNotNull(options?.Value, nameof(options));
+    }
+
+    public ValueTask DisposeAsync()
+        => ValueTask.CompletedTask;
+
+    public async IAsyncEnumerator<ReadResult> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+    {
+        IEnumerable<ReadResult> results = Enumerable.Empty<ReadResult>();
+        for (int i = _startIndex; i < _options.Values.Count; i++)
+        {
+            // Identifier has already been validated
+            var identifier = DicomIdentifier.Parse(_options.Values[i]);
+
+            // Attempt to read the data
+            IReadOnlyList<VersionedInstanceIdentifier> instances = identifier.Type switch
+            {
+                ResourceType.Study => await _instanceStore.GetInstanceIdentifiersInStudyAsync(_partition.PartitionKey, identifier.StudyInstanceUid, cancellationToken),
+                ResourceType.Series => await _instanceStore.GetInstanceIdentifiersInSeriesAsync(_partition.PartitionKey, identifier.StudyInstanceUid, identifier.SeriesInstanceUid, cancellationToken),
+                _ => await _instanceStore.GetInstanceIdentifierAsync(_partition.PartitionKey, identifier.StudyInstanceUid, identifier.SeriesInstanceUid, identifier.SopInstanceUid, cancellationToken),
+            };
+
+            if (instances.Count > 0)
```

Comment:
nit: generally like the flow like, instead of if and else
if (invalid case)
{
    throw;
}
valid case

Reply:
Here we actual preserve the error to send to the error log, but I can put the error first as that is more typical